modules/service/src/main/scala/lucuma/odb/service/ProgramService.scala

```diff
@@ -162,13 +179,60 @@ object ProgramService {
           }
 
         // Update proposals. This can fail in a few ways.
-        val updateProposals: F[List[Program.Id]] =
-          SET.proposal.fold(Nil.pure[F]) {
-            proposalService.updateProposals(_)
+        val updateProposals: F[Either[UpdateProgramsError, List[Program.Id]]] =
+          SET.proposal.fold(Nil.asRight.pure[F]) {
+            proposalService.updateProposals(_).map(_.asRight).recover {
+              case ProposalService.ProposalUpdateException.CreationFailed => 
+                UpdateProgramsError.ProposalCreationFailed.asLeft
+              case ProposalService.ProposalUpdateException.InconsistentUpdate =>
+                UpdateProgramsError.ProposalInconsistentUpdate.asLeft
+            }
           }
 
-        // Combie the results
-        (setup >> updatePrograms, updateProposals).mapN(_ |+| _)
+        // A stable identifier (ie. a `val`) is needed for the enums.
+        val enumsVal = enums
+
+        def tagToProposalStatus(tag: Tag): Either[UpdateProgramsError, enumsVal.ProposalStatus] =
+          Enumerated[enumsVal.ProposalStatus].fromTag(tag.value).toRight(UpdateProgramsError.InvalidProposalStatus(tag))
+
+        def userCanChangeProposalStatus(ps: enumsVal.ProposalStatus): Option[Unit] =
```

Comment:
Would it complicate things too much for this to be `Boolean`?

Reply:
Nope. I'm not sure why I did it that way. Probably distracted by other issues. 😄